Implement matrix truncation for latex to enhance the display of large matrices

SamLubelsky opened the pull request:
<!-- Your title above should be a short description of what
was changed. Do not include the issue number in the title. -->

#### References to other Issues or PRs
<!-- If this pull request fixes an issue, write "Fixes #NNNN" in that exact
format, e.g. "Fixes #1234" (see
https://tinyurl.com/auto-closing for more information). Also, please
write a comment on that issue linking back to this pull request once it is
open. -->

Fixes #16251. 


#### Brief description of what is fixed or changed
I fixed the merge conflict for PR#26284 which seems to have been abandoned.  It should auto-merge into master now.

#### Other comments
Original Author: mohamedrezk122

#### Release Notes

<!-- Write the release notes for this release below between the BEGIN and END
statements. The basic format is a bulleted list with the name of the subpackage
and the release note for this PR. For example:

* solvers
  * Added a new solver for logarithmic equations.

* functions
  * Fixed a bug with log of integers. Formerly, `log(-x)` incorrectly gave `-log(x)`.

* physics.units
  * Corrected a semantical error in the conversion between volt and statvolt which
    reported the volt as being larger than the statvolt.

or if no release note(s) should be included use:

NO ENTRY

See https://github.com/sympy/sympy/wiki/Writing-Release-Notes for more
information on how to write release notes. The bot will check your release
notes automatically to see if they are formatted correctly. -->

<!-- BEGIN RELEASE NOTES -->
* printing
  * Large matrices printed in LaTeX are now automatically condensed(can be disabled)

<!-- END RELEASE NOTES -->

elshorbagyx committed:
author: add Mohamed Rezk to .mailmap

elshorbagyx committed:
printing: Implement large matrix truncation for latex for better printing

The `_print_matrix_contents` method now includes options for compactifying
and setting maximum visible rows and columns, providing a more readable
representation. This improvement prevents printing slowdowns for large
matrices and improves overall user experience. This modification shall
solve #16251.

SamLubelsky committed:
moved around the parameters so the merge will go through

SamLubelsky committed:
moved some more stuff to make the merge go through

SamLubelsky committed:
Merge branch 'sympy:master' into pr/26284

oscarbenjamin commented:
This looks good although I am not sure about the option names.

There should be tests for what happens under different settings.

Ideally all printers would have this option and in fact displaying large matrices should not be done by default. I wonder about having something like a symbolic `ShowAll(expr)` to handle this rather than a printer option.

oscarbenjamin commented on sympy/printing/latex.py:
```diff
@@ -3049,6 +3079,15 @@ def latex(expr, **settings):
     max: Integer or None, optional
         Sets the upper bound for the exponent to print floating point numbers in
         fixed-point format.
+    mat_compact: boolean, optional
+        Truncate large Matrix for printing. Default is ``True``, to avoid printing
+        slowdown.
+    mat_visible_rows: int, optional
+        Sets the maximum number of rows to be printed. Default is ``15``, and
+        it doesn't take effect if mat_compact is ``False``.
+    mat_visible_cols: int, optional
+        Sets the maximum number of columns to be printed. Default is ``15``, and
+        it doesn't take effect if mat_compact is ``False``.
```
I don't think that this many matrix specific options are warranted.

Let's start with something basic: we can have an option like `truncate` that defaults to true. It could be used for more than just matrices but for now it could be implemented only for matrices. Ideally other printers like `str/repr` should support the same and it should be controllable via `init_printing`.

SamLubelsky replied:
Ok, I just made the changes for latex printing.  I agree truncation should also work for the other printers and I will try to get that working.  Should I put that in this PR or open a new one?

Also, do you think there is a way to implement this for many of the printers simultaneously(srepr, unicode, etc), or will it have to be a new implementation for each individual printer?

SamLubelsky committed:
changed truncation to make parameters more clear, added tests.

renamed mat_compact to truncate, removed visible_rows
and visible_columns parameters.  Added tests for each
truncatable case.

SamLubelsky committed:
Merge branch 'sympy:master' into pr/26284

SamLubelsky committed:
Merge branch 'pr/26284' of https://github.com/SamLubelsky/sympy into pr/26284

oscarbenjamin commented on sympy/printing/latex.py:
```diff
@@ -166,6 +166,7 @@ class LatexPrinter(Printer):
         "parenthesize_super": True,
         "min": None,
         "max": None,
+        "mat_compact": True,
```
I still don't think `mat_compact` is the right kind of option to add. Rather there should be a general option for shortening all possible outputs (not just matrices). Its purpose should be to avoid producing absurdly large output in general even if for now it is only applied to matrices.

SamLubelsky replied:
I agree, just forgot to change mat_compact to truncate in a few spots.  The docstring specifies that truncate currently only truncates matrix but might be extended to truncate other things in the future.

SamLubelsky committed:
fixed some old instances of mat_compact where truncate
should now be.

oscarbenjamin commented on sympy/printing/latex.py:
```diff
@@ -166,6 +166,7 @@ class LatexPrinter(Printer):
         "parenthesize_super": True,
         "min": None,
         "max": None,
+        "truncate": True,
```
Using `truncate=False` expresses what is wanted in a negated sense. Maybe what would be better is the opposite like `full_output=True` (default is `False`) or something like that...

SamLubelsky committed:
changed truncate to full_output.

This naming makes more sense if we plan to add a lot
more truncating operations to sympy, beyond just matrices.
